h2/src/main/org/h2/command/dml/Insert.java

```diff
@@ -381,13 +394,17 @@ private boolean handleOnDuplicate(DbException de) {
 
         ArrayList<String> variableNames = new ArrayList<>(
                 duplicateKeyAssignmentMap.size());
-        Expression[] row = list.get(getCurrentRowNumber() - 1);
+        Expression[] row = (currentRow == null) ? list.get(getCurrentRowNumber() - 1) 
+        		: new Expression[columns.length];
         for (int i = 0; i < columns.length; i++) {
             String key = table.getSchema().getName() + "." +
                     table.getName() + "." + columns[i].getName();
             variableNames.add(key);
+            if (currentRow != null) {
+            	row[i] = ValueExpression.get(currentRow[i]);
+            }
             session.setVariable(key,
-                    row[i].getValue(session));
+            		(currentRow == null) ? row[i].getValue(session) : currentRow[i]);
```

Comment:
The same problem with indentation in lines 404 and 407.

Also I think that this piece of code can be written with only one branch for better readability, something like
```Java
            Value value;
            if (currentRow != null) {
                value = currentRow[i];
                row[i] = ValueExpression.get(value);
            } else {
                value = row[i].getValue(session);
            }
            session.setVariable(key, value);
```

Reply:
done